File: leetcode/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        d = {'(': ')', '[': ']', '{': '}'}
        for c in s:
            if c in d:
                stack.append(c)
            elif stack and d[stack[-1]] == c:
                stack.pop()
            else:
                return False
        return len(stack) == 0

File: leetcode/20.c
bool isValid(char * s){
    int i, j = 0;
    for(i = 0; s[i]; i++)
    {
        if(s[i] == '(' || s[i] == '{' || s[i] == '[') {
            s[j] = s[i];
            j ++;
        } else if(j) {
            j --;
            if((s[j] == '(' && s[i] != ')') || \
               (s[j] == '{' && s[i] != '}') || \
               (s[j] == '[' && s[i] != ']')) {
                return false;
            }
        } else {
            return false;
        }
    }
    return j == 0;
}
